### PR TITLE
[Issue 3127][python-client] Replace Exceptions with PulsarExceptions

### LINKED ISSUE #3127: Python Pulsar client appears to throw regular `Exception`s instead of `PulsarException`s
#### Expected behavior
The Python client should throw a subclass of `Exception`, not just `Exception`.

#### Actual behavior
An Exception is thrown.  See, for example:

```
In [2]: p = pulsar.Client()

In [3]: c = p.get_consumer("foo")
2018-12-04 17:24:48.860 INFO  ConnectionPool:63 | Created connection for pulsar://localhost:6650
2018-12-04 17:24:48.871 ERROR ClientConnection:329 | [<none> -> pulsar://localhost:6650] Failed to establish connection: Connection refused
2018-12-04 17:24:48.872 INFO  ClientConnection:1237 | [<none> -> pulsar://localhost:6650] Connection closed
2018-12-04 17:24:48.872 ERROR ClientImpl:374 | Error Checking/Getting Partition Metadata while Subscribing- 5
2018-12-04 17:24:48.872 INFO  ClientConnection:195 | [<none> -> pulsar://localhost:6650] Destroyed connection

< deleted internal wrapper code >

Exception: Pulsar error: ConnectError
```

To demonstrate it's not a subclass:

```
In [8]: try:
   ...:     c.receive(timeout_millis=500)
   ...: except Exception as e:
   ...:     print(type(e))
   ...:
<class 'Exception'>
```
#### Steps to reproduce
Easiest way is probably supply a bogus URL to get a ConnectError

#### System configuration
**Pulsar version**: 2.2.0 for both Pulsar and the Python client

### PR DESCRIPTION
**Pull request by lbenc135**
Fixes #7600 

### Motivation

As the issue says, the Python client throws `Exception`s instead of a subclass of it (`PulsarException`), so the clients must catch the blanket Exception.

### Modifications

Every C `PulsarException` is now thrown in Python with the same type.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
- Extended Python tests to check for PulsarExceptions

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: don't know

### Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by lbenc135** [comment]
Hi @sijie @merlimat, could you please take a look?

**Comment by n0nvme** [comment]
Is it possible to re-run all tests? Logs have expired and it's impossible to see what failed:(

**Comment by aahmed-se** [comment]
@n0nvme rebase and resolve all conflicts first , then push those changes the tests will rerun.

**Comment by lbenc135** [comment]
@n0nvme @aahmed-se I rebased and improved the PR a bit. Some tests are failing, although they seem flaky to me. Let me know if you have any comments.

**Review by tuteng — APPROVED** [review]

**Comment by tuteng** [comment]
@BewareMyPower PTAL

File: pulsar-client-cpp/python/pulsar_test.py [hunk 1]
```diff
@@ -147,11 +159,8 @@ def test_producer_consumer(self):
         self.assertTrue(msg)
         self.assertEqual(msg.data(), b'hello')
 
-        try:
-            msg = consumer.receive(100)
-            self.assertTrue(False)  # Should not reach this point
-        except:
-            pass  # Exception is expected
+        with self.assertRaises(PulsarException):
```
Comment (merlimat) [inline pulsar-client-cpp/python/pulsar_test.py:162]: In these tests, could we assert for `pulsar.Timeout` exception?

**Review by merlimat — APPROVED** [review]

File: pulsar-client-cpp/python/pulsar_test.py (hunk 1 above)
Comment (lbenc135) [reply pulsar-client-cpp/python/pulsar_test.py]: Yes, refactored

**Review by BewareMyPower — APPROVED** [review]

**Comment by BewareMyPower** [comment]
@merlimat @tuteng @jiazhai Could we merge this PR currently?

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**